lib/sudo-system/src/wait.rs

```diff
@@ -0,0 +1,158 @@
+use std::io;
+
+use libc::{
+    c_int, WCONTINUED, WEXITSTATUS, WIFCONTINUED, WIFEXITED, WIFSIGNALED, WIFSTOPPED, WNOHANG,
+    WSTOPSIG, WTERMSIG, WUNTRACED, __WALL,
+};
+use sudo_cutils::cerr;
+
+use crate::{interface::ProcessId, signal::SignalNumber};
+
+/// Wait for a process to change state.
+///
+/// Calling this function will block until a child specified by [`WaitPid`] has changed state. This
+/// can be configured further using [`WaitOptions`].
+pub fn waitpid<P: Into<WaitPid>>(
+    pid: P,
+    options: WaitOptions,
+) -> Result<(ProcessId, WaitStatus), WaitError> {
+    let pid = pid.into().into_pid();
+    let mut status: c_int = 0;
+
+    let pid =
+        cerr(unsafe { libc::waitpid(pid, &mut status, options.flags) }).map_err(WaitError::Io)?;
+
+    if pid == 0 && options.flags & WNOHANG != 0 {
+        return Err(WaitError::NotReady);
+    }
+
+    Ok((pid, WaitStatus { status }))
+}
+
+/// Error values returned when [`waitpid`] fails.
+#[derive(Debug)]
+pub enum WaitError {
+    // No children were in a waitable state.
+    //
+    // This is only returned if the [`WaitOptions::no_hang`] option is used.
+    NotReady,
+    // Regular I/O error.
+    Io(io::Error),
+}
+
+/// Which child process to wait for.
+pub enum WaitPid {
+    /// Wait for any child process.
+    Any,
+    /// Wait for the child process with the given ID.
+    One(ProcessId),
+}
+
+impl WaitPid {
+    fn into_pid(self) -> ProcessId {
+        match self {
+            Self::Any => -1,
+            Self::One(pid) => pid,
+        }
+    }
+}
+
+impl From<ProcessId> for WaitPid {
+    fn from(pid: ProcessId) -> Self {
+        debug_assert!(pid > 0);
```

Comment:
Yeah that'd be nice because using `-1` for us most likely means that one of our `system` functions returned an invalid PID that should be mapped to something else.